modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+      var result = _getRef(i);
+      _collapse(i);
+      _leave();
+      return result;
+    }
+
+    /*
+      Clear the contents of this List.
+
+      .. warning::
+
+        Clearing the contents of this List will invalidate all existing
+        references to the elements contained in this List.
+    */
+    proc clear() {
+      _enter();
+
+      // Manually call destructors on each currently allocated element.
+      for i in 1.._size do {
+        ref item = _getRef(i);
+        _destroy(item);
+      }
+
+      if _blocks != nil then
+        for i in 1.._blocks.capacity do {
+          ref block = _blocks[i];
+          if block == nil then
+            continue;
+          _totalCapacity -= block.capacity;
+          delete block;
+          block = nil;
+        }
+
+      if _blocks != nil then
+        delete _blocks;
+
+      _sanity(_totalCapacity == 0);
+      _blocks = nil;
+      _size = 0;
+
+      _leave();
+    }
+
+    /*
+      Return a zero-based index into this List of the first item whose value
+      is equal to x.
+
+      :arg x: An element to search for.
+      :arg start: The start index to start searching from.
+      :arg end: The end index to stop searching at.
+
+      :return: The index of the element to search for.
+
+      :throws IllegalArgumentError: If the given element cannot be found.
+    */
+    proc indexOf(x: eltType, start: int=1, end: int=size): int throws {
+      _enter();
+
+      try _boundsCheckLeaveOnThrow(start);
+      try _boundsCheckLeaveOnThrow(end);
+
+      for i in start..end do
+        if x == _getRef(i) then {
+          _leave();
+          return i;
+        }
+
+      _leave();
+
+      const msg = "No such element in List: " + x:string;
+      throw new owned
+        IllegalArgumentError(msg);
+
+      // Should never reach here.
+      return -1;
+    }
+
+    /*
+      Return the number of times a given element is found in this List.
+
+      :arg x: An element to count.
+
+      :return: The number of times a given element is found in this List.
+      :rtype: `int`
+    */
+    proc count(x: eltType): int {
+      _enter();
+
+      var result = 0;
+
+      for i in 1.._size do {
+        if x == _getRef(i) then
+          result += 1;
+      }
+
+      _leave();
+      return result;
+    }
+
+    /*
+       Sort the elements of this List in place using their default sort order.
+
+       .. warning::
+
+        Sorting the contents of this List may invalidate existing references
+        to the elements contained in this List.
+    */
+    proc sort() {
+      // TODO: Need to look in the Sort module to see the API.
+    }
+
+    /*
+      Sort the items of this List in place using a comparator.
+
+      .. warning::
+
+        Sorting the elements of this List may invalidate existing references
+        to the elements contained in this List.
+
+      :arg comparator: A comparator object used to sort this List.
+    */
+    proc sort(comparator) {
```

Comment:
I think we'll need to work on the sort module to get this to work directly on list.

You can and should make comparator have a default value of Sort.defaultComparator.

For now, you could call toArray, clear the list, sort the array, and then append the sorted array to the list. I'd leave a comment that says making this copy is not ideal.

Reply:
That's definitely not ideal, but that should work.